hypershift-operator/controllers/scheduler/dedicated_content_serving_nodes.go

```diff
@@ -0,0 +1,204 @@
+	}
+	log.Info("Reconciling node")
+
+	if _, hasServingComponentLabel := node.Labels[hyperv1.RequestServingComponentLabel]; !hasServingComponentLabel {
+		log.Info("Skipping node because it doesn't have the control plane serving component label")
+		return ctrl.Result{}, nil
+	}
+
+	if _, hasHostedClusterLabel := node.Labels[hyperv1.HostedClusterLabel]; !hasHostedClusterLabel {
+		log.Info("Skipping node because it has not been allocated to a hosted cluster")
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("Node has been allocated to a hosted cluster, checking whether hosted cluster still exists.")
+	name := node.Labels[HostedClusterNameLabel]
+	namespace := node.Labels[HostedClusterNamespaceLabel]
+	hc := &hyperv1.HostedCluster{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: namespace, Name: name}, hc); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, fmt.Errorf("failed to get hosted cluster %s/%s: %w", namespace, name, err)
+		}
+		log.Info("The hosted cluster is not found. Deleting node.")
+		if err := r.Delete(ctx, node); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to delete node: %w", err)
+		}
+		log.Info("Node deleted")
+	} else {
+		log.Info("The hosted cluster exists, will check again later.")
+		return ctrl.Result{RequeueAfter: 1 * time.Minute}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+type DedicatedServingComponentScheduler struct {
+	client.Client
+	createOrUpdate upsert.CreateOrUpdateFN
+}
+
+func (r *DedicatedServingComponentScheduler) SetupWithManager(mgr ctrl.Manager, createOrUpdateProvider upsert.CreateOrUpdateProvider) error {
+
+	r.createOrUpdate = createOrUpdateProvider.CreateOrUpdate
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&hyperv1.HostedCluster{}).
+		WithOptions(controller.Options{
+			RateLimiter:             workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 10*time.Second),
+			MaxConcurrentReconciles: 10,
+		}).Named("DedicatedServingComponentScheduler")
+	return builder.Complete(r)
+}
+
+func (r *DedicatedServingComponentScheduler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	hcluster := &hyperv1.HostedCluster{}
+	log := ctrl.LoggerFrom(ctx, "controller", "DedicatedServingComponentScheduler", "hostedcluster", req.NamespacedName.String())
+	err := r.Get(ctx, req.NamespacedName, hcluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("hostedcluster not found, aborting reconcile", "name", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get cluster %q: %w", req.NamespacedName, err)
+	}
+	if !hcluster.DeletionTimestamp.IsZero() {
+		log.Info("hostedcluster is deleted, nothing to do")
+		return ctrl.Result{}, nil
+	}
+	if hcTopology := hcluster.Annotations[hyperv1.TopologyAnnotation]; hcTopology != hyperv1.DedicatedRequestServingComponentsTopology {
+		log.Info("hostedcluster does not use isolated request serving components, nothing to do")
```

Comment:
nit: for term consistency should this file be named dedicated_request_serving_nodes?